### PR TITLE
Promote the provenance field in status

### PR DESCRIPTION
**Pull request by chuangw6**
# Changes

/kind feature

Fixes https://github.com/tektoncd/pipeline/issues/6309.

Prior, the `provenance` field in status was an alpha feature gated by
the dedicated feature flag named `enable-provenance-in-status` which is
disabled by default.

Now, we move this feature to beta by enabling this feature flag by default
with the installation of Tekton Pipelines. In future, we will completely
remove this feature flag once we feel this feature is stable.

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)

# Release Notes

```release-note
Promote `provenance` field to beta by setting the existing feature flag `enable-provenance-in-status` to be true by default with the installation of Tekton Pipeline. This feature flag will be completely removed once we consider this as a stable feature. That said, users can choose to opt out this by setting this feature flag to false. 
```

### CODE REVIEW

**Comment by jerop** [comment]
@chuangw6 this PR needs a rebase, added it to v0.48 milestone

**Comment by chuangw6** [comment]
cc @chitrangpatel PTAL. Thanks!

File: test/larger_results_sidecar_logs_test.go [hunk 1]
```diff
@@ -114,6 +124,12 @@ func TestLargerResultsSidecarLogs(t *testing.T) {
 					t.Fatalf(`The expected taskrun does not match created taskrun. Here is the diff: %v`, d)
 				}
 			}
+
+			// reset result extraction method feature flag
+			if err := updateConfigMap(ctx, c.KubeClient, system.Namespace(), config.GetFeatureFlagsConfigName(), map[string]string{"results-from": previousResultExtractionMethod}); err != nil {
+				t.Fatal(err)
+			}
+
```
Comment (chitrangpatel) [inline test/larger_results_sidecar_logs_test.go:132]: Should we wrap this in a function and we can defer it? That way, if there is an error, the config will be reset so that it does not interfere with tests later.

**Review by chitrangpatel — CHANGES_REQUESTED** [review]
Just one comment. Otherwise lgtm.

File: test/larger_results_sidecar_logs_test.go (hunk 1 above)
Comment (chuangw6) [reply test/larger_results_sidecar_logs_test.go]: Good point! Done. PTAL!

**Review by chitrangpatel — APPROVED** [review]

File: pkg/apis/config/store.go [hunk 2]
```diff
@@ -49,16 +49,12 @@ func FromContextOrDefaults(ctx context.Context) *Config {
...
-	defaults, _ := NewDefaultsFromMap(map[string]string{})
-	featureFlags, _ := NewFeatureFlagsFromMap(map[string]string{})
-	metrics, _ := newMetricsFromMap(map[string]string{})
-	spireconfig, _ := NewSpireConfigFromMap(map[string]string{})
 
 	return &Config{
-		Defaults:     defaults,
-		FeatureFlags: featureFlags,
-		Metrics:      metrics,
-		SpireConfig:  spireconfig,
```
Comment (jerop) [inline pkg/apis/config/store.go:61]: this seems like an unrelated cleanup, is that right? can it go in a separate commit and pull request?

File: pkg/reconciler/pipelinerun/pipelinerun_test.go [hunk 3]
```diff
@@ -8870,6 +8878,9 @@ spec:
 			if err != nil {
 				t.Fatalf("Got an error getting reconciled run out of fake client: %s", err)
 			}
+			tt.expectedPipelineRun.Status.Provenance = &v1beta1.Provenance{
+				FeatureFlags: alphaFeatureFlags,
+			}
```
Comment (jerop) [inline pkg/reconciler/pipelinerun/pipelinerun_test.go:8883]: is there a reason for doing this instead of adding it directly in the expectedpipelinerun yaml?

adding it to the yaml ensures that there's only one place to check for the expected pipelinerun status

File: pkg/reconciler/taskrun/taskrun_test.go [hunk 4]
```diff
@@ -1697,6 +1697,20 @@ status:
...
+	failedOnReconcileFailureTaskRun.Status.Provenance = &v1beta1.Provenance{
+		FeatureFlags: featureFlags,
+	}
+	failedOnReconcileFailureTaskRun.Status.RetriesStatus[0].Provenance = &v1beta1.Provenance{
+		FeatureFlags: featureFlags,
+	}
+
+	retriedTaskRun.Status.Provenance = &v1beta1.Provenance{
+		FeatureFlags: config.DefaultFeatureFlags.DeepCopy(),
+	}
```
Comment (jerop) [inline pkg/reconciler/taskrun/taskrun_test.go:1713]: why not change the yamls?

**Review by jerop — COMMENTED** [review]
thanks @chuangw6! generally looks good, left some comments and questions

File: pkg/apis/config/store.go (hunk 2 above)
Comment (chuangw6) [reply pkg/apis/config/store.go:61]: Yes it's sort of unrelated, but really motivated by this pr - where we need to get the default configmap at many places.

Happy to separate it into another pr.

File: pkg/reconciler/pipelinerun/pipelinerun_test.go (hunk 3 above)
Comment (chuangw6) [reply pkg/reconciler/pipelinerun/pipelinerun_test.go]: Thanks for the feedback. Done.

No particular for unit tests, but we do need to use variables in e2e tests instead of adding it in yamls.

File: pkg/reconciler/taskrun/taskrun_test.go (hunk 4 above)
Comment (chuangw6) [reply pkg/reconciler/taskrun/taskrun_test.go]: Done. Adding it into yamls directly. PTAL. Thank you!

File: pkg/reconciler/pipelinerun/pipelinerun.go [hunk 5]
```diff
@@ -1224,19 +1224,17 @@ func storePipelineSpecAndMergeMeta(ctx context.Context, pr *v1beta1.PipelineRun,
...
-			pr.Status.Provenance.RefSource = meta.RefSource
-		}
-		if meta != nil && meta.RefSource != nil && pr.Status.Provenance.ConfigSource == nil {
-			pr.Status.Provenance.ConfigSource = (*v1beta1.ConfigSource)(meta.RefSource)
-		}
+	if meta != nil && meta.RefSource != nil && pr.Status.Provenance.RefSource == nil {
+		pr.Status.Provenance.RefSource = meta.RefSource
+	}
+	if meta != nil && meta.RefSource != nil && pr.Status.Provenance.ConfigSource == nil {
+		pr.Status.Provenance.ConfigSource = (*v1beta1.ConfigSource)(meta.RefSource)
```
Comment (jerop) [inline pkg/reconciler/pipelinerun/pipelinerun.go:1237]: I thought this feature was being promoted to beta so it should be guarded behind the alpha feature flag? 🤔

**Review by jerop — CHANGES_REQUESTED** [review]
Please add more documentation for the provenance field

It's just one sentence in: https://github.com/tektoncd/pipeline/blob/main/docs/pipelineruns.md

May be a reference to how it is populated, used in https://github.com/tektoncd/chains/tree/main/pkg/chains/formats/slsa/v2

And a note that it's now a beta feature -- also we should add it to the beta features table

File: pkg/apis/config/store.go (hunk 2 above)
Comment (jerop) [reply pkg/apis/config/store.go]: please open another PR with the clean up and I'll prioritize reviewing it 🙏🏾

Comment (chuangw6) [reply pkg/apis/config/store.go]: SG https://github.com/tektoncd/pipeline/pull/6639. PTAL. Thanks!

File: pkg/reconciler/pipelinerun/pipelinerun.go (hunk 5 above)
Comment (chuangw6) [reply pkg/reconciler/pipelinerun/pipelinerun.go:1237]: This feature was only gated by its dedicated feature flag per https://github.com/tektoncd/pipeline/issues/5632.

File: docs/additional-configs.md [hunk 6]
```diff
@@ -284,7 +280,6 @@ Features currently in "alpha" are:
...
 | [Object Params and Results](pipelineruns.md#specifying-parameters)                                  | [TEP-0075](https://github.com/tektoncd/community/blob/main/teps/0075-object-param-and-result-types.md)                     | [v0.38.0](https://github.com/tektoncd/pipeline/releases/tag/v0.38.0) |                               |                             |
 | [Trusted Resources](./trusted-resources.md)                                                         | [TEP-0091](https://github.com/tektoncd/community/blob/main/teps/0091-trusted-resources.md)                                 | N/A                                                                  | `trusted-resources-verification-no-match-policy`  |
-| [`Provenance` field in Status](pipeline-api.md#provenance)                                          | [issue#5550](https://github.com/tektoncd/pipeline/issues/5550)                                                             | N/A                                                                  | `enable-provenance-in-status` |
```
[the remark this answers is not in the document]
Comment (chuangw6) [reply docs/additional-configs.md:287]: Yes, thanks for catching that!

**Comment by chuangw6** [comment]
Thank you @jerop for the review. I've addressed all the comments. Please take a look.

 Regarding the feature flag, we will keep the dedicated feature flag but enable it by default. please see the https://github.com/tektoncd/pipeline/issues/6309#issuecomment-1553458310 for details.

**Review by jerop — APPROVED** [review]
Thank you @chuangw6!

/unhold

File: docs/additional-configs.md [hunk 7]
```diff
@@ -245,10 +245,11 @@ Defaults to "ignore".
 
 - `results-from`: set this flag to "termination-message" to use the container's termination message to fetch results from. This is the default method of extracting results. Set it to "sidecar-logs" to enable use of a results sidecar logs to extract results instead of termination message.
 
-- `enable-provenance-in-status`: set this flag to "true" to enable recording
+- `enable-provenance-in-status`: Set this flag to `true` to enable populating
```
Comment (jerop) [inline docs/additional-configs.md:248]:
```suggestion
- `enable-provenance-in-status`: Set this flag to `"true"` to enable populating
```

File: docs/additional-configs.md [hunk 8]
```diff
@@ -245,10 +245,11 @@ Defaults to "ignore".
...
   the `provenance` field in `TaskRun` and `PipelineRun` status. The `provenance`
   field contains metadata about resources used in the TaskRun/PipelineRun such as the
-  source from where a remote Task/Pipeline definition was fetched.
+  source from where a remote Task/Pipeline definition was fetched. By default, this is set to `true`.
+  To disable populating this field, set this flag to `false`.
```
Comment (jerop) [inline docs/additional-configs.md:252]:
```suggestion
  To disable populating this field, set this flag to `"false"`.
```

File: docs/additional-configs.md (hunk 7 above)
Comment (chuangw6) [reply docs/additional-configs.md]: Done. Thank you @jerop

File: docs/additional-configs.md (hunk 8 above)
Comment (chuangw6) [reply docs/additional-configs.md]: Done. Thank you!
